matron/src/device/device_midi.c

```diff
@@ -86,91 +86,298 @@ int dev_midi_init(void *self, unsigned int port_index, bool multiport_device) {
+            event = NULL;
+        }
+
+        /* Voice category message? */
+        if (c < 0xF0) {
+            parser->channel = c & 0x0F;
+            parser->status = c & 0xF0;
+
+            /* The event consumes x bytes of data... (subtract 1 for the status byte) */
+            parser->nr_bytes_total = midi_event_length(parser->status) - 1;
+
+             /* 0  bytes read so far */
+            parser->nr_bytes = 0;
+        }
+        else if (c == MIDI_SYSEX) {
+            parser->status = MIDI_SYSEX;
+            parser->nr_bytes = 0;
+        }
+        else {
+            parser->status = 0;    /* Discard other system messages (0xF1-0xF7) */
+        }
+
+        return event; /* Return SYSEX event or NULL */
+    }
+
+    /* Data/parameter byte */
+
+    /* Discard data bytes for events we don't care about */
+    if(parser->status == 0) {
+        return NULL;
+    }
+
+    /* Max data size exceeded? (SYSEX messages only really) */
+    if(parser->nr_bytes == MIDI_PARSER_MAX_DATA_SIZE) {
+        parser->status = 0; /* Discard the rest of the message */
+        return NULL;
+    }
+
+    /* Store next byte */
+    parser->data[parser->nr_bytes++] = c;
+
+    /* Do we still need more data to get this event complete? */
+    if(parser->status == MIDI_SYSEX || parser->nr_bytes < parser->nr_bytes_total) {
+        return NULL;
+    }
+
+    /* Event is complete, return it.
+     * Running status byte MIDI feature is also handled here. */
+    parser->event.type = parser->status;
+    parser->event.channel = parser->channel;
+    parser->nr_bytes = 0; /* Reset data size, in case there are additional running status messages */
+
+    switch(parser->status) {
+    case NOTE_OFF:
+    case NOTE_ON:
+    case KEY_PRESSURE:
+    case CONTROL_CHANGE:
+    case PROGRAM_CHANGE:
+    case CHANNEL_PRESSURE:
+        parser->event.param1 = parser->data[0]; /* For example key number */
+        parser->event.param2 = parser->data[1]; /* For example velocity */
+        break;
+
+    case PITCH_BEND:
+        /* Pitch-bend is transmitted with 14-bit precision. */
+        parser->event.param1 = (parser->data[1] << 7) | parser->data[0];
+        break;
+
+    default: /* Unlikely */
+        return NULL;
+    }
+
+    return &parser->event;
+}
+
 void *dev_midi_start(void *self) {
     struct dev_midi *midi = (struct dev_midi *)self;
     union event_data *ev;
 
+    midi_parser_t *parser = midi_parser_new();
+    midi_event_t *evt;
+
     ssize_t read = 0;
-    uint8_t byte = 0;
-    uint8_t msg_buf[256];
-    uint8_t msg_pos = 0;
-    uint8_t msg_len = 0;
+
+    int npfds;
+    struct pollfd *pfds;
+    npfds = 1 + snd_rawmidi_poll_descriptors_count(midi->handle_in);
+    pfds = alloca(npfds * sizeof(struct pollfd));
+    snd_rawmidi_poll_descriptors(midi->handle_in, &pfds[1], npfds - 1);
 
     do {
-        read = snd_rawmidi_read(midi->handle_in, &byte, 1);
-
-        if (byte >= 0xf8) {
-            clock_midi_handle_message(byte);
```

Comment:
The original midi parse called `clock_midi_handle_message(...)` while the new parser does not. I have not investigated deeply but my guess is that this change will break the existing midi clock sync.

Reply:
good catch. I will add it back.